src/plugins/orangehrmRecruitmentPlugin/Controller/ApplicantController.php

```diff
@@ -0,0 +1,317 @@
+    }
+
+    /**
+     * @param CandidateVacancy $candidateVacancy
+     * @param int $applicantId
+     * @param string $status
+     * @param int $vacancyId
+     */
+    private function setApplicantVacancy(
+        CandidateVacancy $candidateVacancy,
+        int $applicantId,
+        string $status,
+        int $vacancyId
+    ): void {
+        $candidateVacancy->getDecorator()->setCandidateById($applicantId);
+        $candidateVacancy->getDecorator()->setVacancyById($vacancyId);
+        $candidateVacancy->setStatus($status);
+        $candidateVacancy->setAppliedDate($this->getDateTimeHelper()->getNow());
+    }
+
+    /**
+     * @param CandidateHistory $applicantHistory
+     * @param int $applicantId
+     * @param int $action
+     * @return void
+     */
+    private function setCommonApplicantHistoryAttributes(
+        CandidateHistory $applicantHistory,
+        int $applicantId,
+        int $action
+    ): void {
+        $applicantHistory->getDecorator()->setCandidateById($applicantId);
+        $applicantHistory->setAction($action);
+        $applicantHistory->setPerformedDate($this->getDateTimeHelper()->getNow());
+    }
+
+
+    /**
+     * @return ValidationDecorator
+     */
+    protected function getValidationDecorator(): ValidationDecorator
+    {
+        if (!$this->validationDecorator instanceof ValidationDecorator) {
+            $this->validationDecorator = new ValidationDecorator();
+        }
+        return $this->validationDecorator;
+    }
+
+    /**
+     * @return ParamRuleCollection|null
+     */
+    protected function getParamRuleCollection(): ?ParamRuleCollection
+    {
+        return new ParamRuleCollection(
+            new ParamRule(
+                self::PARAMETER_RESUME,
+                new Rule(Rules::BASE_64_ATTACHMENT)
+            ),
+            new ParamRule(
+                self::PARAMETER_FIRST_NAME,
+                new Rule(Rules::STRING_TYPE)
+            ),
+            $this->getValidationDecorator()->notRequiredParamRule(
+                new ParamRule(
+                    self::PARAMETER_MIDDLE_NAME,
+                    new Rule(Rules::STRING_TYPE)
+                ),
+                true
+            ),
+            new ParamRule(
+                self::PARAMETER_LAST_NAME,
+                new Rule(Rules::STRING_TYPE)
+            ),
+            new ParamRule(
+                self::PARAMETER_EMAIL,
+                new Rule(Rules::EMAIL)
+            ),
+            $this->getValidationDecorator()->notRequiredParamRule(
+                new ParamRule(
+                    self::PARAMETER_CONTACT_NUMBER,
+                    new Rule(Rules::PHONE)
+                )
+            ),
+            $this->getValidationDecorator()->notRequiredParamRule(
+                new ParamRule(
+                    self::PARAMETER_VACANCY_ID,
+                    new Rule(Rules::POSITIVE)
+                )
+            ),
```

Comment:
```suggestion
            new ParamRule(
                self::PARAMETER_VACANCY_ID,
                new Rule(Rules::POSITIVE)
            ),
```
:thinking: I think this is required

Reply:
oh,, my bad. Done